kafka-instrumentation/src/main/java/dev/responsive/kafka/clients/ResponsiveKafkaClientSupplier.java

```diff
@@ -0,0 +1,276 @@
+ * rely on coarse-grained locks around sections reading/writing shared state.
+ */
+public final class ResponsiveKafkaClientSupplier implements KafkaClientSupplier, Configurable {
+  private static final Logger LOGGER = LoggerFactory.getLogger(ResponsiveKafkaClientSupplier.class);
+  private final KafkaClientSupplier wrapped;
+  private final Factories factories;
+  private boolean eosV2 = false;
+  private boolean configured = false;
+  private Metrics metrics;
+  private EndOffsetsPoller endOffsetsPoller;
+
+  private final Map<String, ReferenceCounted<MetricPublishingCommitListener>> metricListeners
+      = new HashMap<>();
+
+  public ResponsiveKafkaClientSupplier() {
+    this(new Factories() {}, new DefaultKafkaClientSupplier());
+  }
+
+  ResponsiveKafkaClientSupplier(final Factories factories, final KafkaClientSupplier wrapped) {
+    this.factories = factories;
+    this.wrapped = wrapped;
+  }
+
+  @Override
+  public void configure(final Map<String, ?> configs) {
+    LOGGER.info(
+        "Configuring the client supplier. Got configs: {}",
+        configs.entrySet().stream()
+            .map(e -> e.getKey() + ":" + e.getValue())
+            .collect(Collectors.joining("\n"))
+    );
+    eosV2 = (configs.get(StreamsConfig.PROCESSING_GUARANTEE_CONFIG))
+        .equals(StreamsConfig.EXACTLY_ONCE_V2);
+    final JmxReporter jmxReporter = new JmxReporter();
+    jmxReporter.configure(configs);
+    final MetricsContext metricsContext
+        = new KafkaMetricsContext("dev.responsive", new HashMap<>());
+    metrics = factories.createMetrics(
+        new MetricConfig(),
+        Collections.singletonList(jmxReporter),
+        Time.SYSTEM,
+        metricsContext
+    );
+    endOffsetsPoller = factories.createEndOffsetPoller(configs, metrics);
+    configured = true;
+  }
+
+  @Override
+  public Admin getAdmin(final Map<String, Object> config) {
+    return wrapped.getAdmin(config);
+  }
+
+  @Override
+  public Producer<byte[], byte[]> getProducer(final Map<String, Object> config) {
+    assert configured;
+    if (!eosV2) {
+      return wrapped.getProducer(config);
+    }
+    LOGGER.info("Creating responsive producer");
+    final String tid = threadIdFromProducerConfig(config);
+    synchronized (this) {
+      return factories.createResponsiveProducer(
+          wrapped.getProducer(config),
+          Collections.unmodifiableList(
+              Arrays.asList(
+                  addOrRefCommitListener(tid),
+                  new CloseListener(tid)
+              )
+          )
+      );
+    }
+  }
+
+  @Override
+  public Consumer<byte[], byte[]> getConsumer(final Map<String, Object> config) {
+    assert configured;
+    if (!eosV2) {
+      return wrapped.getConsumer(config);
+    }
+    LOGGER.info("Creating responsive consumer");
+    final String tid = threadIdFromConsumerConfig(config);
+    synchronized (this) {
+      // TODO: the end offsets poller call is kind of heavy for a synchronized block
```

Comment:
Yeah why not just do that up front/outside this lock?

Reply:
I think this is fine after my last patch